test/puppetlabs/puppetdb/command_test.clj

```diff
@@ -172,45 +172,44 @@
 (defn make-cmd
   "Create a command pre-loaded with `n` attempts"
   [n]
-  {:command nil :version nil :annotations {:attempts (repeat n {})}})
+  {:headers {:id "foo-id-1"
+             :received (tfmt/unparse (tfmt/formatters :date-time) (now))}
+   :body (json/generate-string  {:command "some catalog" :payload "foo" :version 10 :annotations {:attempts (repeat n {})}})})
 
 (deftest command-retry-handler
-  (testing "Retry handler"
-    (with-redefs [metrics.meters/mark!  (call-counter)
-                  mql/annotate-with-attempt (call-counter)]
-
-      (testing "should log errors"
-        (let [publish  (call-counter)]
-          (testing "to DEBUG for initial retries"
-            (let [log-output (atom [])]
-              (binding [*logger-factory* (atom-logger log-output)]
-                (mql/handle-command-retry (make-cmd 1) nil publish))
-
-              (is (= (get-in @log-output [0 1]) :debug))))
-
-          (testing "to ERROR for later retries"
-            (let [log-output (atom [])]
-              (binding [*logger-factory* (atom-logger log-output)]
-                (mql/handle-command-retry (make-cmd mql/maximum-allowable-retries) nil publish))
-
-              (is (= (get-in @log-output [0 1]) :error)))))))))
-
-(deftest test-error-with-stacktrace
-  (with-redefs [metrics.meters/mark!  (call-counter)]
-    (let [publish  (call-counter)]
-      (testing "Exception with stacktrace, no more retries"
-        (let [log-output (atom [])]
-          (binding [*logger-factory* (atom-logger log-output)]
-            (mql/handle-command-retry (make-cmd 1) (RuntimeException. "foo") publish))
-          (is (= (get-in @log-output [0 1]) :debug))
-          (is (instance? Exception (get-in @log-output [0 2])))))
-
-      (testing "Exception with stacktrace, no more retries"
-        (let [log-output (atom [])]
-          (binding [*logger-factory* (atom-logger log-output)]
-            (mql/handle-command-retry (make-cmd mql/maximum-allowable-retries) (RuntimeException. "foo") publish))
+  (testing "Should log retries as debug for less than 4 attempts"
```

Comment:
This is an example of how we could test this stuff going forward. Below doesn't use the test message handler macros but is able to isolate the logic on when to retry, from what happens when a message is retried.